### PR TITLE
[NT-489] Default card selection improvement

### PR DESCRIPTION
**Pull request by justinswart**
# 📲 What

Fixes an issue where the first stored card was always selected in any context where the payment method could be set even when it was not used to back the project.

# 🤔 Why

Our logic for handling this was a bit shortsighted in that it would simply select the first card in the list every time. Instead we need to confirm that the card was in fact used to back the project and hasn't since been removed.

# 🛠 How

Updated our logic for this to take into account the `PaymentSource` of the `Backing` and if it's set, check that the card is still an available payment source.

# ✅ Acceptance criteria

- [ ] Back a project and delete the payment method used to back the project. When navigating to change your payment source, there shouldn't be any payment source selected by default and the `Confirm` button should be disabled.
- [ ] Back a project using Apple Pay. Similarly there shouldn't be any default payment method selected when changing.

### CODE REVIEW

File: Library/ViewModels/PledgeCreditCardViewModel.swift [hunk 1]
```diff
@@ -96,16 +96,20 @@ public final class PledgeCreditCardViewModel: PledgeCreditCardViewModelInputs,
       .takeWhen(Signal.merge(cardConfiguredAsSelected, self.selectButtonTappedProperty.signal))
       .map { $0.id }
 
-    let cardIsSelected = cardAndSelectedCard
-      .map(==)
+    let cardIsSelected = Signal.merge(
+      creditCard.mapConst(false),
```
Comment (justinswart) [inline Library/ViewModels/PledgeCreditCardViewModel.swift]: Emitting `false` here to ensure that the card has a title even though we're not setting a selected card.

File: Library/ViewModels/PledgePaymentMethodsViewModel.swift [hunk 2]
```diff
@@ -136,18 +133,32 @@ public final class PledgePaymentMethodsViewModel: PledgePaymentMethodsViewModelT
...
+) -> ([PledgeCreditCardViewData], GraphUserCreditCard.CreditCard?) {
+  let data = cards.compactMap { card -> PledgeCreditCardViewData? in
     guard let cardBrand = card.type?.rawValue else { return nil }
 
     let isAvailableCardType = availableCardTypes.contains(cardBrand)
 
     return (card, isAvailableCardType, project.location.displayableName)
   }
+
+  guard let backing = project.personalization.backing else {
```
Comment (justinswart) [inline Library/ViewModels/PledgePaymentMethodsViewModel.swift]: If this project hasn't been backed then we do just select the first card.

File: Library/ViewModels/PledgePaymentMethodsViewModel.swift [hunk 3]
```diff
@@ -136,18 +133,32 @@ public final class PledgePaymentMethodsViewModel: PledgePaymentMethodsViewModelT
...
+
+  var backedCard: GraphUserCreditCard.CreditCard?
+
+  let cardIds = cards.map { $0.id }
+
+  if let cardId = backing.paymentSource?.id, cardIds.contains(cardId) {
+    backedCard = cards.first
+  }
+
+  return (data, backedCard)
```
Comment (Scollaco) [inline Library/ViewModels/PledgePaymentMethodsViewModel.swift]: Here I think we could simply use 
```
let backedCard = cards.first(where: { $0.id == backing.paymentSource?.id })
```

File: Library/ViewModels/PledgePaymentMethodsViewModelTests.swift [hunk 4]
```diff
@@ -265,6 +265,55 @@ final class PledgePaymentMethodsViewModelTests: TestCase {
     }
   }
 
+  func testReloadPaymentMethods_LoggedIn_ApplePayCapable_isTrue_BackedCardRemoved() {
+    let filteredCards = GraphUserCreditCard.template.storedCards.nodes
+      .filter { $0.id != GraphUserCreditCard.visa.id
+      }
```
Comment (Scollaco) [inline Library/ViewModels/PledgePaymentMethodsViewModelTests.swift]: Could you put the `}` on the same line?

File: Library/ViewModels/PledgePaymentMethodsViewModelTests.swift [hunk 5]
```diff
@@ -265,6 +265,55 @@ final class PledgePaymentMethodsViewModelTests: TestCase {
...
+
+      self.vm.inputs.configureWith((User.template, project, true))
+      self.vm.inputs.viewDidLoad()
+
+      self.scheduler.run()
+
+      self.applePayButtonHidden.assertValues([false])
+      self.reloadPaymentMethodsCards.assertValue(response.me.storedCards.nodes)
+      self.reloadPaymentMethodsAvailableCardTypes.assertValues([
+        [true, true, true, true, true, false, true]
```
Comment (Scollaco) [inline Library/ViewModels/PledgePaymentMethodsViewModelTests.swift]: Why there's an unavailable card? 🤔

File: Library/ViewModels/PledgePaymentMethodsViewModel.swift (hunk 3 above)
Comment (justinswart) [reply Library/ViewModels/PledgePaymentMethodsViewModel.swift]: lol thanks that's much better

File: Library/ViewModels/PledgePaymentMethodsViewModelTests.swift (hunk 5 above)
Comment (justinswart) [reply Library/ViewModels/PledgePaymentMethodsViewModelTests.swift]: This is just the Project template's available card types, I copied this test from the one above it.

Comment (Scollaco) [reply Library/ViewModels/PledgePaymentMethodsViewModelTests.swift]: Oh cool!

**Pull request merged**